### ISSUE TITLE
Security: Please Re-Instate WebRTC Blocking Immediately.

### ISSUE THREAD

**Issue by scramblr** [uBlockOrigin/uBlock-issues#1799]
### Prerequisites

- [X] I verified that this is not a filter issue (MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues))
- [X] This is not a [support issue or a question](https://old.reddit.com/r/uBlockOrigin/)
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue
- [X] The issue is not present after wholly disabling uBlock Origin ("uBO") in the browser
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### I tried to reproduce the issue when...

- [X] uBO is the only extension
- [X] uBO with default lists/settings
- [X] using a new, unmodified browser profile

### Description

Back in September, a thread was created (https://github.com/uBlockOrigin/uBlock-issues/issues/1723) and WebRTC was removed from the uBlock codebase. This has silently exposed many people who rely on airtight configurations to never expose their external IP address. While the idea of masking your external IP address was mocked in the thread, I would politely refer you to the use cases of Tor, Proxies, and other IP masking technologies. WebRTC is used to hop over these masks and expose a user's true external IP address. 

It was a very important safety tool, and it was removed without even alerting anyone. Please add it back. 

### A specific URL where the issue occurs

https://browserleaks.com/webrtc

### Steps to Reproduce

This is N/A.

### Expected behavior

Please re-instate WebRTC protections immediately. 

### Actual behavior

N/A

### uBlock Origin version

All future versions

### Browser name and version

Chrome, Chromium, etc

### Operating System and version

All

**Comment by sickcodes** [comment]
Why was this removed? WebRTC IP Leaking via STUN requests is textbook tracking

**Comment by opsxcq** [comment]
This issue affects everyone, please fix it ASAP.

**Comment by uBlock-user** [comment]
> While the idea of masking your external IP address was mocked in the thread, I would politely refer you to the use cases of Tor, Proxies, and other IP masking technologies. WebRTC is used to hop over these masks and expose a user's true external IP address.

WebRTC settings oly blocked leaking local router IP address, your external IP address was never hidden in the first place.

**Comment by scramblr** [comment]
You don't seem to understand how this works.

**Comment by uBlock-user** [comment]
There's nothing to understand, "Prevent WebRTC from leaking local IP addresses" setting speaks for itself. Your external IP address was never hidden, adding the setting back won't hide it either.

**Comment by MLGRadish** [comment]
It seems like the name is confusing. Renaming the feature to "disable WebRTC" should fix the issue.

**Comment by uBlock-user** [comment]
> Renaming the feature to "disable WebRTC" should fix the issue.

https://github.com/gorhill/uBlock/issues/2853

Doubt that's possible in Chromium based browsers.

**Comment by MLGRadish** [comment]
How about a Firefox only feature then?

**Comment by scramblr** [comment]
Can you please explain to me why the uBlock version from ublock.org is able to protect my IP from exposure still, then? https://pbs.twimg.com/media/FDdOCBUVcAMFrjG?format=jpg&name=4096x4096

> There's nothing to understand, "Prevent WebRTC from leaking local IP addresses" setting speaks for itself. Your external IP address was never hidden, adding the setting back won't hide it either.

**Comment by duaneking** [comment]
Is it possible that this was bought out and so this was removed as a way to make it less effective for users and thus more valuable for the new owners?

**Comment by uBlock-user** [comment]
> Can you please explain to me why the uBlock version from ublock.org is able to protect my IP from exposure still, then? https://pbs.twimg.com/media/FDdOCBUVcAMFrjG?format=jpg&name=4096x4096

Side-effect of that setting on your "airtight" configuration perhaps. Hiding external IP address has never been the intension of the WebRTC setting.

**Comment by scramblr** [comment]
You are completely dismissive and condescending about all of this, and you don't have a grasp on what you're doing. Enjoy your day.

**Comment by nameless1337** [comment]
> > Can you please explain to me why the uBlock version from ublock.org is able to protect my IP from exposure still, then? https://pbs.twimg.com/media/FDdOCBUVcAMFrjG?format=jpg&name=4096x4096
> 
> Side-effect of that setting on your "airtight" configuration perhaps. Hiding external IP address has never been the intension of the WebRTC setting.
> 
> Your external IP address is already sent to the website when you navigate to it via a document request. So in the end it's still exposed.

You really don't get it do you

**Comment by gorhill** [comment]
The aptly named _"Prevent WebRTC from leaking local IP addresses"_ never had for purpose to hide your public-facing IP address, this is silly to think so -- it's literally in the name of the setting, and uBO is not a VPN or proxy service. Now with regard to non-public-facing IP addresses, the setting is no longer needed in modern browsers:

LAN IP addresses are not disclosed when using my ISP:
![Screenshot from 2021-11-05 18-59-05](https://user-images.githubusercontent.com/585534/140587911-1a668fcd-4c92-4a8e-99ec-386edb577ac1.png)

LAN IP addresses and ISP IP address are not disclosed when using a VPN:
![Screenshot from 2021-11-05 19-00-28](https://user-images.githubusercontent.com/585534/140587824-06a879ab-a990-416c-b215-16f364e3be82.png)

Your public-facing IP address is visible to all sites you visit, it's silly to think that if it's not reported at `https://browserleaks.com/webrtc` specifically it means the site `browserleaks.com` can't see it, you literally requested a web page from that site, they know your public-facing IP address, they sent the HTTP response to it, see for yourself: <https://browserleaks.com/ip>.

**Comment by gorhill** [comment]
So it appears the invalid issue filed here is a result of [this tweet](https://twitter.com/notdan/status/1456706246170472450). The tweet contains many falsehoods:

![Screenshot from 2021-11-06 06-41-01](https://user-images.githubusercontent.com/585534/140607241-d54aadf7-ac9f-4d72-8957-a303b7557257.png)

### "silently removed"

False. There is [an issue describing the motivation for the change](https://github.com/uBlockOrigin/uBlock-issues/issues/1723), and it's in [the release notes](https://github.com/gorhill/uBlock/releases/tag/1.38.0).

If you do not trust uBO that much, then it's on you to closely follow development and read release notes -- your failure to do so does not make documented changes "silent".

### "removed all WebRTC blocking"

False. uBO has never ever supported "WebRTC blocking". The setting was called _"Prevent WebRTC from leaking **local** IP addresses"_, and its sole purpose was to prevent **local** IP address leakage through WebRTC, [as reported here](https://threatpost.com/webrtc-found-leaking-local-ip-addresses/110803/) (2015).

This WebRTC flaw has been fixed natively in Firefox and Chromium [more than two years ago](https://github.com/uBlockOrigin/uBlock-issues/issues/1723#issuecomment-919903359), making the uBO setting _"Prevent WebRTC from leaking **local** IP addresses"_ obsolete ever since this was fixed in those browsers.

The setting is still present and honored in Firefox for Android because it was not fixed natively as in the desktop version of Firefox. If you disable the setting in uBO in Firefox for Android and visit <https://browserleaks.com/webrtc>, you will likely see the _"Local IP Address"_ entry being filled with one or more IP addresses. This is what the uBO setting is meant to prevent, and it works as expected once you enable it in that browser.

### "making MANY users vulnerable to IP unmasking"

False. As seen in [my comment above](https://github.com/uBlockOrigin/uBlock-issues/issues/1799#issuecomment-962273710), there is no "IP unmasking", and OP has **failed** to demonstrate this, and only provided a link to <https://browserleaks.com/webrtc>, apparently expecting that _"Public IP Address"_ should be _n/a_ -- no actual steps or technical details are provided to make the case there is an actual issue, so essentially just [FUD](https://en.wikipedia.org/wiki/Fear%2C_uncertainty%2C_and_doubt).

As explained in [my comment above](https://github.com/uBlockOrigin/uBlock-issues/issues/1799#issuecomment-962273710), this is silly, your public IP address is by definition visible to all sites you visit -- sites don't need WebRTC for this, just visit [this page](https://browserleaks.com/ip) on the same site, it reports your public-facing IP address regardless of whether WebRTC is enabled or not.

The tweet has two screenshots, devoid of any details about the context. The `browserleaks.com`-related screenshot in the tweet shows that there was no _"Local IP address"_ leakage, thus thereby confirming that the _"Prevent WebRTC from leaking **local** IP addresses"_ setting is indeed no longer needed.

**Comment by uBlock-user** [comment]
OP and his posse who downvoted my responses are the same people spreading that mis-information via tweets on some ignorant belief that their ISP associated IP address is somehow blocked via that setting. It's astounding as to how far they will go because of the setting's removal for desktop browsers.

**Comment by gorhill** [comment]
At this point I am convinced all this is completely in bad faith -- hence why he refuses to come up with details regarding his claims.

Consider that [he posted](https://twitter.com/notdan/status/1456718997831839746) a screenshot of sleazy uBlock supposedly blocking WebRTC as if natively without mentioning that this can be accomplish only after one add the filter `*$webrtc`.

There is no setting in the dashboard of sleazy uBlock to control WebRTC behavior, this can only done through adding a filter.

**You can accomplish just the same with uBO through it's own filter syntax:** `*##+js(nowebrtc)`.

[uBO's implementation](https://github.com/uBlockOrigin/uAssets/commit/ed5f4062f40911b5257b05be69068656ba609058) (Jun 2016) predates [that of sleazy uBlock](https://github.com/gorhill/uBlock/commit/f5ab54efb66866c79d1776fc9998f66d097334d9) (Oct 2018), and in both cases the purpose is to address sites using WebRTC to push ads, not to hide public-facing IP address through WebRTC.

**Comment by uBlock-user** [comment]
OP and his posse are trolling, ignore them. I don't see anyone with a sensible mind joining in the conversation in the tweets.

**Comment by gorhill** [comment]
Trolling or not, it's someone with a fair number of followers (many whose opinions appear captive and accepted without questioning that the claim is valid) who spread [FUD](https://en.wikipedia.org/wiki/Fear,_uncertainty,_and_doubt) about uBO:

> FUD is generally a strategy to influence perception by disseminating negative and dubious or false information and a manifestation of the appeal to fear. 

This is exactly what this case is about, and such case of FUD directed at uBO deserves to be thoroughly documented -- though we will likely never know what ultimately motivated this person to try to undermine the trust people have in uBO -- trusted earned over years of never compromising users' best interests and through refusing to sell out.

Summary:
- Spreads FUD: <https://twitter.com/notdan/status/1456706246170472450>
    - My answer: <https://twitter.com/gorhill/status/1456776940493451268>
    - More detailed answer: <https://github.com/uBlockOrigin/uBlock-issues/issues/1799#issuecomment-962437677>
    - People can't reproduce: <https://twitter.com/fernacen/status/1457003371060137988>
    - Refuses to substantiate his claims and conveniently disengage when asked
    to substantiate:
        - <https://twitter.com/notdan/status/1456978262802849794>
        - <https://twitter.com/gorhill/status/1457005713146978309>
- Calls for his followers to mob the repo: <https://twitter.com/notdan/status/1456709677396336647>
- Misleads people: <https://twitter.com/notdan/status/1456718997831839746>
    - My answer: <https://twitter.com/gorhill/status/1457059084029550592>
    - A more detailed answer: <https://github.com/uBlockOrigin/uBlock-issues/issues/1799#issuecomment-962496931>

If anything this shows that people need to not unquestionably accept someone claims **especially** when they refuses to substantiate them outright and/or through deflection, regardless of how popular is the source<sup>[1]</sup>.

---

[1]  This is why personally I typically provides upfront as much verifiable information when I make claims, I want people to be able to compare their own findings with mines -- I don't want them to blindly trust me.

**Comment by uBlock-user** [comment]
You won't be able to change their minds no matter how much you try to convince them, that's the impression I got at this thread, so document as needed, but engaging them in fruitful conversation is simply never going to happen. Talking to these group kind of people is like talking to a wall.

**Comment by gorhill** [comment]
> You won't be able to change their minds

I am not trying to achieve this, I just wanted to document it in details for posterity. Done.

**Comment by gorhill** [comment]
> won't be able to change their minds

I am not trying to achieve this, I just wanted to document this in details to save time for whoever may want to look more into all this and for future reference if this happens again.

**Comment by uBlock-user** [comment]
What puzzles me is why they pick uBO only, what they want is not offered by any content/ad blocker in existence, so in theory they should go after other blockers like ABP/AdGuard/Brave too.

**Comment by gwarser** [comment]
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">This leaked my internal IPs (all of them) on latest Chrome with ublock both debian and windows when mic/camera permission is granted. <a href="https://t.co/9jExpx0UIo">https://t.co/9jExpx0UIo</a></p>&mdash; PFFaith (@Phhaith1) <a href="https://twitter.com/Phhaith1/status/1475277275947945985?ref_src=twsrc%5Etfw">December 27, 2021</a></blockquote>


**"when mic/camera permission is granted."**

I believe this is true. I read about this somewhere - cannot find now. But why someone will give browser permission to connect for only purpose to share IP / fingerprinting? On the other hand, with WebRTC blocked, connections will not be possible at all, even legit ones. 

@gorhill

**Comment by gorhill** [comment]
On my side, when not selecting mic/camera permission, no local IP addresses are leaking. When selecting mic/camera permission, there is an exception -- I don't have a mic/camera.

> I believe this is true

Did you reproduce?

**Comment by gwarser** [comment]
I too don't have mic or camera. Arkenfox issue is linking to convincing documents.

**Comment by gorhill** [comment]
By the way, I am now blocked by that <https://twitter.com/notdan> person ([ref](https://user-images.githubusercontent.com/585534/147666758-7d5362c7-c84d-419c-b4fe-655f1763b435.png)), because I "dared" to provide a link to the thread here ([ref](https://user-images.githubusercontent.com/585534/147666871-5a063fe0-63dc-42ff-bf86-2b0061062fc1.png)) in response to his recent tweet ([ref](https://user-images.githubusercontent.com/585534/147666933-a1d77021-6ff8-44bc-a4b1-accf4c05b504.png)) in which he referred to the thread here without (conveniently) linking to it.

I consider the advice of using `*$+js(nowebrtc)` (which curiously he was unable to figure out on his own while he had no issue figuring sleazy uBlock's `*$webrtc`) to not be ideal, at least it should come with a disclosure that unconditionally patching the DOM this way on all sites makes one stand out since the patching can be seen by page's JS, and thus used as fingerprinting information (through `Object.getOwnPropertyDescriptor(window, 'RTCPeerConnection')`).
 
The core issue here is the persistent unwillingness to provide exact repro steps which can be analyzed and acted upon if needed. Asking me to _"re-Instate WebRTC blocking immediately"_ based on two screenshots devoid of context and where attempts to understand what is being reported are being stonewalled (running a pottymouth on twitter is not a substitute for competent technical information to help reproduce an claimed issue), while those who cared to try to reproduce found the opposite of the original claim.

Ultimately, the whole issue here has been debased by a bad faith person, and if there is an actual issue to solve, this should be opened as a separate issue with all the actual relevant technical information as required -- all this is not specific to uBO, this is how any project work -- issues have to be substantiated to be acted upon.

**Comment by gorhill** [comment]
> I too don't have mic or camera. Arkenfox issue is linking to convincing documents.

The original comment was linking to `https://browserleaks.com/webrtc`. On that page, there is a _"Audio Capture Permissions"_ which I am able to activate ( I do have a microphone), and I do not see any local IP addresses leaked as a result.

**Comment by uBlock-user** [comment]
> This has silently exposed many people who rely on airtight configurations to never expose their external IP address. 

OP's reasoning for adding "WebRTC immediately", which is why this thread was going to go nowhere and was doomed from the beginning.

**Comment by gwarser** [comment]
I borrowed camera, and on `https://browserleaks.com/webrtc` when permissions are given **permanently** (probably page script limitation - they may detect on reload when temporary permission timeouts) it does detect my LAN IP:

![image](https://user-images.githubusercontent.com/886325/147690645-9029ea61-940c-4651-9e5b-0463c59ce839.png)


On `https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/` even temporary permission allow to list my LAN IP:

![image](https://user-images.githubusercontent.com/886325/147690934-513f60b3-a4bd-4113-99c3-193b6422d429.png)

Firefox 97.0a1 20211229092739
uBO 1.40.3b0


Chrome does not seem to support temporary permissions, granting them allow both pages to read LAN IP.

Chrome 96.0.4664.110 (Official Build) (64-bit) with custom build uBO 1.40.3b0

**Comment by gorhill** [comment]
Thanks for reporting this. Question:

- Did you test with a pre-1.38.0 version to find out if you get the same result?
- Does this also leak your ISP address when behind of VPN?
- Dos this happen with the microphone permission? (I wasn't able to reproduce in that case)

**Comment by uBlock-user** [comment]
@u-RraaLL

**Comment by u-RraaLL** [comment]
I agree with @gwarser 's findings: Allowing either microphone or camera on browserleaks "permanently" leaks LAN IP.
In my case - 2 routers = 2 LAN IPs.
On `webrtc.github.io`, allowing even temporarily leaks both LAN IPs.

Dropping down to uBO 1.37.2 and enabling "Prevent WebRTC from leaking local IP addresses" fixes the issue on both sites.

There is no public IP leakage from behind the system-wide VPN I'm using.
Haven't tried any VPN extensions though.

**Comment by gorhill** [comment]
> There is no ISP IP leakage from behind the system-wide VPN I'm using.

Thanks, I was waiting for a confirmation.

To be clear, these findings have nothing to do with the opening comment, the two screenshots at the root of why the issue was opened had no local IP address in it.

In any case, I don't consider the new finding is a good reason to bring back the setting, because:

- browsers block camera/microphone access by default and a user must expressly grants a permission to access in the first place
    - a reminder that the purpose of the original setting was to prevent all sites from being able to silently access private IP address information, including ISP one when behind a VPN
- the granted permission applies to one site only -- the site one expressly **trusted** to access microphone/camera feed
    - both Firefox and Chromium allows to block by default such that no request to access is shown at any time (default is to ask)
- it would be utterly silly to worry about a given site being able to know your local IP address **after** you have expressly given that site permission to listen in on your conversations and access the feed of your camera
- uBO taking over the `webRTCIPHandlingPolicy` created actual conflicts with extension-based VPN/proxies, and even in browser-based VPN ([Opera for example](https://github.com/uBlockOrigin/uBlock-issues/issues/1297)), and possibly causing users to be _less_ protected as a result
    - Typically, good extension-based VPN/proxies offer a way to tighten `webRTCIPHandlingPolicy`, so one should be using these.

**Comment by duaneking** [comment]
With respect:

* Not all browsers block this access by default.  It's gotten better, but it's not universal, and many brands actively work to minimize device security because they see the default installed software and lack of security as a revenue opportunity.  We all know about junkware, let's not pretend it does not exist.

* Granted perms are often difficult to track or revoke in the moment, and advertisers basically make a living finding ways to get past your security/defenses and actively work to limit the security available in browsers because they see browser security as a competitive threat to their continued business.  Lets not pretend that companies like Google do not exist that make billions every year from advertising and collection of user data.

* Just because I want to view a websites text content does not mean I want to give it access to my mic or camera constantly.

* Just because a website thinks it's been give access to a camera device does not mean that device is not intentionally virtualized and secured due to lack of trust. 

* Just because I want to give one specific page or process access to a devices mic or camera for a single specific thing, does not mean I want that same device to be used in a way that I do not want.

* Access on browsers for mics/camera is utterly broken, not fine grain enough, and should not be used as an excuse to not make the rest of the system more secure.

**Comment by gorhill** [comment]
I suggest you just install an extension that let you manage WebRTC, for example the official one from webrtc.org, <https://chrome.google.com/webstore/detail/webrtc-network-limiter/npeicpdbkakmehahjeeohfdhnlpdklia>. In Firefox, this can be done in `about:config`. As already stated, that setting in uBO has been no longer relevant in uBO for years, and was actually causing issues for extension-based VPN/proxy.

**Comment by Agych** [comment]
Wanted initially to leave this comment in #1826.

@gorhill , according to the release notes:

"The setting "Prevent WebRTC from leaking local IP addresses" has been removed since it is no longer necessary in modern browsers"

But what about people who are using non-latest browsers?
There are people using FF 52/56, Chromium-87-based browsers who still need Flash, latest Cent and 360 browsers are running Chromium 86 engine, also Opera 58 where the UI is still not as bad as it is in 60+, XP-users are sticking to {Chromium 48, Opera 36, the mentioned above 360}, latest FF versions have horrible UI, and etc etc etc, plenty examples where (and reasons why) people are consciously using non-latest browsers.

Why removing the setting instead of  keeping it and letting the user decide to use it or not?

I would like to request to bringing this setting back and keep the choice for the user.

**Comment by uBlock-user** [comment]
> There are people using FF 52/56 because of the extensions

Minimum supported version now is Firefox 60.

**Comment by gorhill** [comment]
> keep the choice for the user

Those who are using outdated browser versions have the choice to use outdated uBO versions.
